### DISCUSSION TITLE
How do I get the "change settings per layer" option to show up?

### DISCUSSION THREAD

**Discussion by williamsikkema** [sn4k3/UVtools#519] [Q&A]
I have a resin that requires very long exposure times, and I'm making modifications to it, so I need to routinely test the exposure parameters.  The current exposure time finder generates individual images, not a waterfall, but I can use layer arithmetic to add the appropriate images together.  I need the stage not to move between different exposures at the same z height, and I have found the setting where I can check that box, but it doesn't seem to work for me.  The printer still moves between exposures when it shouldn't.   My printer is a Saturn.  The box that appears to allow me to turn off z motions appears and disappears with the "Change settings per a layer range" option visibility under "tools" -> "edit print parameters".  How do I get the "change settings per layer" option to show up?

**Comment by sn4k3** [comment]
> I need the stage not to move between different exposures at the same z height

It's impossible with the current firmware and printer. So far that is only possible with the freedom of gcode printers.
CTB and others requires a obligatory lift! If you set to 0 firmware will cap to a min/max value and then it can never be 0! That's why UVtools set to 0.1mm on the test, still some printers have a secondary lift on firmware (Firmware TSMC) and that make it go higher than 0.1mm anyway.

**Reply by williamsikkema** [reply]
The Saturn is on this list, though: https://github.com/sn4k3/UVtools/wiki/Printer-compability-with-per-layer-settings-and-advanced-tools

What printers are gcode compatible?

**Reply by williamsikkema** [reply]
Also, how do I get the "change settings per layer" option to show up, even if my printer can't do 0 lifts?  I think I would be ok with very small lifts.  Chitubox allows 0.001 mm lifts, but your software only allows a minimum of 0.01 mm lifts. 

Also, yours has a cap on exposure time of 1000 seconds. Could you remove that upper limit?

**Reply by williamsikkema** [reply]
I can remove the secondary lift with this gcode file run before. I have no idea where the authors of this gcode got the parameters from, as they're not on the "Saturn machine parameters list", or anywhere else

M8016 D0		; The number of milliseconds to wait after z rises
M8015 P0		; Z slow rise speed
M8016 P0		; Z speed of rapid rise and fall
M8008 I5		        ; Acceleration - higher acceleration = faster the overall print time. Too high will cause step loss
M8070 S0		; Peel move SLOW MOVE in mm - 0 = Disabled > speed set by M8015 
M8070 Z0		; Peel lift distance FAST MOVE in mm 0 = Disabled > speed set by M8016
M8510			; Make the configuration effective immediately, but do not save the configuration

**Reply by williamsikkema** [reply]
I could also have each file be a single layer and print different files at different z heights as long as I can prevent homing.  Is there a way to do that?

**Reply by sn4k3** [reply]
> The Saturn is on this list, though: https://github.com/sn4k3/UVtools/wiki/Printer-compability-with-per-layer-settings-and-advanced-tools

It is compatible with all tools within UVtools, but that doesn't mean you are free to do everything you want, there still constrains on firmware.

> What printers are gcode compatible?

There aren't much... NovaMaker printers, IBEE and **Uniformation GKone** which is what i recommend atm if you want gcode printer.

> Also, how do I get the "change settings per layer" option to show up, even if my printer can't do 0 lifts? I think I would be ok with very small lifts. Chitubox allows 0.001 mm lifts, but your software only allows a minimum of 0.01 mm lifts.

That option is only avaliable IF your file format supports it. You need at least CTBv3 or v4 for that option to show. Also required for multiple exposure test.
Never use 0.001mm! you are asking for problems, keep a safe margin and use 0.1mm.

> Also, yours has a cap on exposure time of 1000 seconds. Could you remove that upper limit?

1000 seconds is already a overkill, in which situation you would need more than 1000s? A large exposure time can damage your LCD quickly.

> I can remove the secondary lift with this gcode file run before. I have no idea where the authors of this gcode got the parameters from, as they're not on the "Saturn machine parameters list", or anywhere else

You can try it but make a backup of your actual settings.
Also read this: https://github.com/sn4k3/UVtools/wiki/Rest-times-and-TSMC

> I could also have each file be a single layer and print different files at different z heights as long as I can prevent homing. Is there a way to do that?

No.
But why the current multiple exposure test don't work for you?

**Reply by williamsikkema** [reply]
Thanks for your response!

The resin I'm producing is a hydrogel, and needs exposures in the 500 seconds - 5000 seconds range.  I'm ok with killing the LCD routinely - it's a consumable.

The current multiple exposure test doesn't work for me because the resin is weak and flexible when only partially cured, so movements of partially cured resin destroy it.  The way your multiple exposures work it exposes each section by itself, but when I'm doing long exposures * 10 different samples, it takes a very long time, and I would like to nest them (ie - the 100 second can run simultaneously with the 1000 second, just turn off that portion of the screen after 100 seconds)

This project is very valuable to me, and if you can help me solve it, I'll gladly through a $500 donation your way.

**Reply by sn4k3** [reply]
> just turn off that portion of the screen after 100 seconds

This is only possible with GCode printers. 
Your and most printers require a new layer, and new layer always require a positive lift by firmware requirement. If you can disable second lift and issue a very small lift it may work.
But for your case i think you need to go with the Scripting tool. That way to can control everything and put whatever time you want and auto merge slices. After script is made all you need is run it to repeat the actions.

**Reply by williamsikkema** [reply]
Ok, I figured out a super hacky way to do it. I run that custom gcode to get ride of default lifts.  Then I include a black slice at the end of every waterfall that is set to a z height of 10 mm.  It's not doing any lift to get there, but it goes there to print the black layer, then goes back down to start the print of the next waterfall.

One thing that would be really helpful for me is if I could run multiple lines of layer arithmetic at the same time. :)

**Reply by williamsikkema** [reply]
I can print a file like this, but I can't reimport that kind of file back into UV tools, because it complains about non-monotonic layer heights. :)

**Reply by sn4k3** [reply]
> One thing that would be really helpful for me is if I could run multiple lines of layer arithmetic at the same time. :)

You need to run a script for that.

> I can print a file like this, but I can't reimport that kind of file back into UV tools, because it complains about non-monotonic layer heights. :)

Do you run the latest version? It allow lower from a higher layer if that layer is empty (all black).

**Reply by sn4k3** [reply]
New version can run multiple Layer arithmetic in one operation, use **;** to start a new arithmetic operation
